--- 
title: "Extra quantities for paper"
author: "Rene Welch"
output:
  bookdown::html_document2:
    toc: false
    theme: spacelab
    fig_caption: true
    code_folding: hide
---

So, there are many quantities to fill in the manuscript, and since we are not going to make something reproducible, this is the best I can do:

```{r setup, include=FALSE, echo=FALSE, eval=TRUE}

library(magrittr)
library(tidyverse)
library(gwasrapidd)

source(here::here("src/gwas_funs_redo.R"))

```

```{r load-scccs-gwas, include=FALSE, echo=FALSE, eval=TRUE}

folder <- "2022_06_29_prepare_imputation"
workdir <- here::here("results", folder, "qs", "gwas_clean")
all_files <- list.files(workdir, full.names = TRUE, pattern = "gwas_stats")

stats1 <- all_files %>%
  stringr::str_subset("covar_bmi_age.vdbp") %>%
  purrr::map(qs::qread) %>%
  bind_rows()


all_stats <- here::here("results", folder, "gwas/stats/SCCS_chr4.acount") %>%
  vroom::vroom() %>%
  dplyr::rename_with(snakecase::to_snake_case) %>%
  dplyr::mutate(maf = alt_cts / obs_ct)

finemap_snps <- here::here("results", "2022_08_03_finemapping", "qs",
      "sccs_vdbp_fwd_step.qs") %>%
  qs::qread() %>%
  dplyr::rename(id = term)

var_stats <- all_stats %>%
  dplyr::right_join(dplyr::select(finemap_snps, "id"), by = "id")

finemap_snps %<>%
  dplyr::inner_join(
    dplyr::select(var_stats, id, maf, obs_ct), by = "id") %>%
  dplyr::mutate(
    pve = pve(estimate, std.error, maf, obs_ct))

# get covariates of the model
folder <- "2022_06_29_prepare_imputation"
sampledir <- here::here("results", folder, "tsv", "samples")

samples <- tibble::tibble(
  file = list.files(sampledir, full.names = TRUE),
  platform = str_remove(basename(file), "_best_call_rates.tsv")) %>%
  dplyr::mutate(
    samples = purrr::map(file, readr::read_tsv, col_names = "iid")) %>%
  dplyr::select(-file)

data_dir <- here::here("extracted_data", "SCCS")
vitd_data <- file.path(data_dir, "vitamind_phenotypes.txt") %>%
  readr::read_tsv()

vitd_data %<>%
  dplyr::rename_all(list(snakecase::to_snake_case)) %>%
  dplyr::mutate(iid = str_c(id_number, id_number, sep = "_"))


samples %<>%
  dplyr::mutate(
    samples = purrr::map(samples, dplyr::left_join, vitd_data, by = "iid")) %>%
  tidyr::unnest(cols = c(samples))

# load pca
pcas <- here::here("results", folder, "ld_prune", "sccs_pca.eigenvec") %>%
  vroom::vroom() %>%
  dplyr::rename_with(snakecase::to_snake_case) %>%
  dplyr::mutate(
    across(where(is.numeric), list(~ . / sd(.)), .names = "{.col}"))


# one phenotype files
pheno <- samples %>%
  dplyr::select(iid, vit_25_ohd, vdbp) %>%
  dplyr::mutate(across(-iid, list(~ log1p(.)), .names = "{.col}")) %>%
  dplyr::select(iid, vdbp)

# 1 covariates file
covar <- samples %>%
  dplyr::select(iid, platform, bmi, enrollment_age, sex) %>%
  dplyr::mutate(
    age = case_when(
      enrollment_age < 50 ~ "A",
      enrollment_age <= 54 ~ "B",
      enrollment_age <= 59 ~ "C",
      enrollment_age <= 64 ~ "D",
      TRUE ~ "E"),
    sex = if_else(sex == "M", 1, 0)) %>%
  dplyr::select(-enrollment_age) %>%
  dplyr::inner_join(pcas, by = "iid")

all_data <- covar %>%
  dplyr::select(-platform) %>%
  dplyr::inner_join(pheno, by = "iid") %>%
  dplyr::select(iid, vdbp, tidyselect::everything())

# get genotype for the finemapped snps

snp_range <- finemap_snps %>%
  dplyr::pull(id) %>%
  stringr::str_split("\\:") %>%
  purrr::map_dbl(~ as.numeric(.[2])) %>%
  range()

library(GenomicRanges)
library(VariantAnnotation)
library(biomaRt)
library(Matrix)


gr <- GenomicRanges::GRanges(
  seqnames = 4,
  ranges = IRanges(start = snp_range[1], end = snp_range[2]))

vcf_files <- here::here("results", "2022_06_29_prepare_imputation",
  "impute_subset") %>%
  list.files(full.names = TRUE, pattern = ".dose.vcf.gz") %>%
  stringr::str_subset(".csi", negate = TRUE) %>%
  stringr::str_subset(".tbi", negate = TRUE)

gmat <- get_genotype(gr, vcf_files, finemap_snps)

gmat %<>%
  as.data.frame() %>%
  tibble::rownames_to_column("iid") %>%
  tibble::as_tibble()

all_data %<>%
  dplyr::inner_join(gmat, by = "iid") %>%
  dplyr::select(-iid)

model <- lm(vdbp ~ ., data = all_data)

joint_model0 <- model %>%
  broom::tidy() %>%
  dplyr::filter(term != "(Intercept)") %>%
  dplyr::filter(! term %in% names(all_data)) %>%
  dplyr::filter(!stringr::str_detect(term, "age")) %>%
  dplyr::rename(id = term) %>%
  dplyr::mutate(id = stringr::str_remove_all(id, "`")) %>%
  dplyr::inner_join(
    finemap_snps %>%
      dplyr::select(id, rsids, maf, obs_ct), by = "id") %>%
  dplyr::mutate(
    pve = pve(estimate, std.error, maf, obs_ct))

rsids <- finemap_snps %>%
  dplyr::select(id, rsids) %>%
  tidyr::unnest(cols = c(rsids)) %>%
  dplyr::filter(refsnp_id != "rs143381474")


```

Number of SNPs in chromosome 4: `r stats1 %>% dplyr::filter(chrom == 4) %>% nrow() %>% scales::comma()`

Beta coeffcients

```{r beta-coeff, include=TRUE, echo=FALSE, eval=TRUE}
gwas_snps <- stats1 %>%
  dplyr::inner_join(rsids, by = "id")

gwas_snps %>%
  dplyr::select(refsnp_id, id, beta) %>%
  dplyr::mutate(
    coeff_beta = exp(beta),
    yield = (exp(beta)) / (exp(0)) - 1
    ) %>%
  knitr::kable()

pred2 <- predict(model, newdata = all_data %>%
  dplyr::filter(`4:72618334:A:C` == 2)) %>%
  mean(na.rm = TRUE)

pred0 <- predict(model, newdata = all_data %>%
  dplyr::filter(`4:72618334:A:C` == 0)) %>%
  mean(na.rm = TRUE)

pred2 / pred0 - 1


```

rs705117 p.value: `r stats1 %>% dplyr::filter(chrom == 4 & pos == 72608115) %>%
pluck("p", 1) %>% scales::scientific()` and the coeff. is `r stats1 %>% dplyr::filter(chrom == 4 & pos == 72608115) %>% pluck("beta", 1) %>% exp`

## locus zoom ranges

```{r range-loci, include=TRUE, echo=FALSE, eval=TRUE}
pval <- 5e-8
window <- 250e3

stats2 <- stats1 %>%
  dplyr::filter(p <= pval)

top_snp <- stats2 %>%
  dplyr::group_by(chrom) %>%
  dplyr::top_n(1, wt = -p) %>%
  dplyr::ungroup()


top_snp %<>%
  dplyr::bind_rows(
    top_snp %>%
      dplyr::select(chrom, pos) %>%
      dplyr::inner_join(
        stats1, by = "chrom", suffix = c("ref", "")) %>%
      dplyr::filter(abs(posref - pos) > window / 2) %>%
      dplyr::select(-posref) %>%
      dplyr::group_by(chrom) %>%
      dplyr::top_n(1, wt = -p))


library(VariantAnnotation)
library(Matrix)
library(snpStats)
library(GenomicRanges)
library(ggbio)
library(EnsDb.Hsapiens.v75)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(BSgenome.Hsapiens.UCSC.hg19)
library(biomaRt)
library(ggrepel)


# get significant snps
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
human <- org.Hs.eg.db 
ensdb <- EnsDb.Hsapiens.v75


top_snp %<>%
  dplyr::mutate(id2 = id) %>%
  tidyr::nest(data = -c(id2)) %>%
  dplyr::mutate(
    gr = purrr::map(data, create_gr),
    gr = purrr::map(gr, IRanges::resize, window, fix = "center"))

top_snp %>%
  dplyr::mutate(
    gr = map_chr(gr, as.character)) %>%
  tidyr::unnest() %>%
  dplyr::select(-a_1, -test, -errcode) %>%
  knitr::kable()

```

```{r locus-zoom1, include=TRUE, echo=FALSE, eval=TRUE}

all_files <- here::here("results/2022_06_29_prepare_imputation/gwas") %>%
  list.files(full.names = TRUE, recursive = TRUE) %>%
  stringr::str_subset("chr4")

summary_stats <- all_files %>%
  stringr::str_subset("bmi_age") %>%
  stringr::str_subset("vdbp") %>%
  vroom::vroom() %>%
  dplyr::rename_with(snakecase::to_snake_case)

summary_stats1 <- summary_stats %>%
  dplyr::filter(pos %in% c(72610021, 72610021, 72613036, 72621674)) %>%
  dplyr::filter(test == "ADD")

summary_stats1 %>%
  dplyr::select(-a_1, -test, -errcode) %>%
  dplyr::mutate(p = scales::scientific(p)) %>%
  knitr::kable()

summary_stats2 <- summary_stats %>%
  dplyr::filter(pos %in% c(72838021)) %>%
  dplyr::filter(test == "ADD")

summary_stats2 %>%
  dplyr::select(-a_1, -test, -errcode) %>%
  dplyr::mutate(p = scales::scientific(p)) %>%
  knitr::kable()


```