--- 
title: "Table 1 UK biobank"
author: "Rene Welch"
output:
  bookdown::html_document2:
    toc: false
    theme: spacelab
    fig_caption: true
    code_folding: hide
---

We are going to gather genotype data and make stratified comparisons by sex.

```{r setup, include=FALSE, echo=FALSE, eval=TRUE}

library(magrittr)
library(tidyverse)
library(gwasrapidd)

source(here::here("src/gwas_funs_redo.R"))

# load summary statistics
stats1 <- ukb_afroc_gwas <-
  here::here("results/2022_08_04_ukbiobank_finemapping/qs",
   "afroc_ukb_stats_filter.qs") %>%
  qs::qread()

ld_snps <- stats1 %>%
  dplyr::filter(
    str_detect(id, regex("^rs4588$")) |
    str_detect(id, "rs1352846")) %>%
  dplyr::filter(chrom == 4)


gr <- GenomicRanges::GRanges(seqnames = "04",
  ranges = IRanges::IRanges(
    start = min(ld_snps$pos) - 100,
    end = max(ld_snps$pos) + 100))


# fine-mapping parameters
pval <- 5e-8
window <- 250e3

stats1 %<>%
  dplyr::filter(p <= pval)

# > stats1 %>% dplyr::pull(pos) %>% range
# [1] 72608364 72681824
# > stats1 %>% dplyr::pull(pos) %>% range %>% diff
# [1] 73460
# > stats1 %>% dplyr::pull(pos) %>% summary
#     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
# 72608364 72617612 72618049 72631222 72637197 72681824 

top_snp <- stats1 %>%
  dplyr::group_by(chrom) %>%
  dplyr::top_n(1, wt = -p) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(
    chrom = str_c("0", chrom))


# top_snp %<>%
#   dplyr::bind_rows(
#     top_snp %>%
#       dplyr::select(chrom, pos) %>%
#       dplyr::inner_join(
#         stats1, by = "chrom", suffix = c("ref", "")) %>%
#       dplyr::filter(abs(posref - pos) > window / 2) %>%
#       dplyr::select(-posref) %>%
#       dplyr::group_by(chrom) %>%
#       dplyr::top_n(1, wt = -p))

# Bioconductor part
library(GenomicRanges)
library(VariantAnnotation)
library(biomaRt)
library(Matrix)

top_snp %<>%
  dplyr::mutate(id2 = id) %>%
  tidyr::nest(data = -c(id2)) %>%
  dplyr::mutate(
    gr = purrr::map(data, create_gr),
    gr = purrr::map(gr, IRanges::resize, window * 2, fix = "center"))

vcf_files <- here::here("results/2022_08_04_ukbiobank_finemapping",
  "vcf/out") %>%
  list.files(full.names = TRUE, pattern = "_range.vcf.gz") %>%
  stringr::str_subset(".csi", negate = TRUE) %>%
  stringr::str_subset(".tbi", negate = TRUE)


get_genotype <- function(gr, vcf_files, stats) {

  chrom <- GenomeInfoDb::seqnames(gr) %>%
    as.character()
  vcf_file <- vcf_files

  svp <- VariantAnnotation::ScanVcfParam(
    geno = "GP", which = gr)

  vcf <- VariantAnnotation::readVcf(vcf_file, "hg19", param = svp)
  rnms <- rownames(vcf)
  rnms_rsids <- stringr::str_split(rnms, ",") %>%
    purrr::map_chr(1)

  stats %<>%
    dplyr::mutate(vcf_id = purrr::map(id, ~ which(. == rnms_rsids))) %>%
    dplyr::filter(purrr::map_dbl(vcf_id, length) > 0) %>%
    dplyr::mutate(rnm = purrr::map_chr(vcf_id, ~ rnms[.[1]]))

  vcf0 <- vcf[stats$rnm, ]
  geno_mat <- geno(vcf0)[["GP"]]
  geno_mat <- as.list(geno_mat)
  out_mat <- purrr::map(geno_mat, which.max)

  attr(out_mat, "dim") <- attr(geno_mat, "dim")
  attr(out_mat, "dimnames") <- attr(geno_mat, "dimnames")

  out_mat <- matrix(unlist(out_mat), ncol = ncol(vcf0), byrow = FALSE)
  rownames(out_mat) <- rownames(vcf0)
  colnames(out_mat) <- colnames(vcf0)


  ## there are some significant snps outside the 250 kbps range around the
  ## top snp
  out_mat <- out_mat - 1
  t(out_mat[stats$rnm, ])

}

top_snp %<>%
  dplyr::mutate(
    geno = purrr::map(gr, get_genotype, vcf_files, stats1))

gg <- get_genotype(gr, vcf_files, ld_snps)

# add covariates and response
folder <- "2021_07_01_ukbiobank_stratified_gwas"
data_dir <- here::here("results", folder, "qs")
uk_data <- file.path(data_dir, "ukbiobank_grant_vars.qs") %>%
  qs::qread()
princomp <- file.path(data_dir, "ukbiobank_princomp.qs") %>%
  qs::qread()
  
princomp %<>%
  dplyr::mutate(
    eid = stringr::str_split(iid, "\\_") %>%
      purrr::map_chr(1) %>%
      as.integer()) %>%
  dplyr::select(-iid)
  
uk_data %<>%
  dplyr::inner_join(princomp, by = "eid")

fwd_step_regression <- function(geno, modeldata) {


  fwd_vars <- modeldata %>%
    dplyr::select(-eid, -vitamin_d) %>%
    names()

  geno_tib <- geno %>%
    as.data.frame() %>%
    tibble::rownames_to_column("eid") %>%
    tibble::as_tibble()

  modeldata %<>%
    dplyr::mutate(eid = as.character(eid))

  modeldata %<>%
    dplyr::inner_join(geno_tib, by = "eid") %>%
    dplyr::rename_all(list(~ str_replace_all(., ":", "\\_")))
  stop_criteria <- FALSE
  all_vars <- modeldata %>%
    dplyr::select(-vitamin_d, -eid) %>%
    names()

  linear_model <- function(var, fwd_vars, modeldata) {

    my_data <- modeldata %>%
      dplyr::select(tidyselect::one_of(c("vitamin_d", fwd_vars, var)))
    model <- lm(log1p(vitamin_d) ~ ., data = my_data)
    broom::tidy(model) %>%
      dplyr::filter(term == glue::glue("{var}"))
  }

  i <- 1
  modeldata %<>%
    dplyr::select(-eid)
  while (length(fwd_vars) < length(all_vars) & !stop_criteria) {

    message("iteration:", i)
    rem_vars <- all_vars[!all_vars %in% fwd_vars]
    message("linear models")
    pvalues <- furrr::future_map_dfr(rem_vars, linear_model, fwd_vars,
      modeldata)

    # select variable
    lowest <- pvalues %>%
      dplyr::top_n(1, -p.value) %>%
      dplyr::pull(term) %>%
      stringr::str_remove_all("`")
    pvalues %<>%
      dplyr::filter(term != glue::glue("{x}", x = lowest))

    stop_criteria <- all(pvalues$p.value > 5e-8)
    sum(pvalues$p.value <= 5e-8)
    fwd_vars <- c(fwd_vars, lowest)
    message(lowest)
    i <- i + 1
  }

  finaldata <- modeldata %>%
    dplyr::select(tidyselect::one_of(c("vitamin_d", fwd_vars)))
  finalmodel <- lm(log(vitamin_d) ~ ., data = finaldata)
  out <- broom::tidy(finalmodel) %>%
    dplyr::mutate(
      term = if_else(stringr::str_detect(term, regex("^pc")),
        term, stringr::str_replace_all(term, "\\_", ":")))

  return(out)

}

all_data <- uk_data %>%
  dplyr::select(eid, vitamin_d, age_bin, sex, bmi, starts_with("pc"))

top_snp %<>%
  dplyr::mutate(
    fwd = purrr::map(geno, fwd_step_regression, all_data))

model_snps <- top_snp %>%
  dplyr::select(id2, fwd) %>%
  tidyr::unnest(cols = c(fwd)) %>%
  dplyr::filter(term != "(Intercept)") %>%
  dplyr::filter(! term %in% names(all_data)) %>%
  dplyr::filter(!stringr::str_detect(term, "age")) %>%
  dplyr::filter(!stringr::str_detect(term, "sex")) %>%
  dplyr::mutate(
    term = stringr::str_remove_all(term, "`"),
    coord = stringr::str_split(term, ":") %>%
      purrr::map_chr(~ stringr::str_c(.[1], .[2], .[2], sep = ":")))



```

```{r, include = TRUE, echo=TRUE, eval=TRUE}

top_snp$fwd[[1]] %>%
  dplyr::filter(p.value <= .05) %>%
  dplyr::mutate(p.value = scales::scientific(p.value)) %>%
  knitr::kable()


```


# Figures

```{r geno, include=TRUE, echo=FALSE, eval=TRUE, fig.width=4, fig.height=3, out.width="50%"}

geno_mat <- top_snp$geno[[1]] %>%
  as.data.frame() %>%
  tibble::as_tibble(rownames = "eid")

pdata <- all_data %>%
  dplyr::mutate(eid = as.character(eid)) %>%
  dplyr::inner_join(geno_mat, by = "eid") %>%
  dplyr::select(vitamin_d, sex, rs4588) %>%
  tidyr::pivot_longer(-(1:2), names_to = "id", values_to = "geno")

pdata$vitamin_d %>% median
pdata$vitamin_d %>% iqr
pdata$vitamin_d %>% log1p() %>% median
pdata$vitamin_d %>% log1p() %>% iqr



pdata %<>%
  dplyr::bind_rows(
    pdata %>%
      dplyr::mutate(sex = "both"))

pdata %<>%
  dplyr::mutate(
    sex = factor(sex, levels = c("both", "male", "female")),
    geno = factor(geno))

pal <- c("black", RColorBrewer::brewer.pal(3, "Set1")[1:2]) %>%
  rlang::set_names(levels(pdata$sex))

pdata %>%
  ggplot(aes(x = geno, y = log1p(vitamin_d), colour = sex)) +
    geom_boxplot() +
    facet_wrap(facets = vars(id), nrow = 2) +
    cowplot::theme_minimal_grid() +
    theme(
      strip.background = element_blank(),
      strip.text = element_text(angle = 0, hjust = 0, size = 14),
      axis.text = element_text(size = 12),
      axis.title = element_text(size = 14),
      legend.text = element_text(size = 12),
      legend.title = element_text(size = 14),
      legend.position = "top") +
    labs(x = "# of effect allele copies", y = "log 25-Hydroxy Vitamin D",
      colour = "") +
    scale_color_manual(values = pal)

```



```{r geno2, include=TRUE, echo=FALSE, eval=TRUE, fig.width=4, fig.height=3.5, out.width="50%"}

geno_mat <- top_snp$geno[[1]] %>%
  as.data.frame() %>%
  tibble::as_tibble(rownames = "eid")

pdata <- all_data %>%
  dplyr::mutate(eid = as.character(eid)) %>%
  dplyr::inner_join(geno_mat, by = "eid") %>%
  dplyr::select(vitamin_d, age_bin, rs4588) %>%
  tidyr::pivot_longer(-(1:2), names_to = "id", values_to = "geno") %>%
  dplyr::rename(age = age_bin)


pdata %<>%
  dplyr::mutate(
    geno = factor(geno))


pdata %>%
  ggplot(aes(x = geno, y = log1p(vitamin_d), colour = age)) +
    geom_boxplot() +
    facet_wrap(facets = vars(id), nrow = 2) +
    cowplot::theme_minimal_grid() +
    theme(
      strip.background = element_blank(),
      strip.text = element_text(angle = 0, hjust = 0, size = 14),
      axis.text = element_text(size = 12),
      axis.title = element_text(size = 14),
      legend.text = element_text(size = 12),
      legend.title = element_text(size = 14),
      legend.position = "top") +
    labs(x = "# of effect allele copies", y = "log 25-Hydroxy Vitamin D",
      colour = "Age") +
    guides(colour = guide_legend(nrow = 2, byrow = TRUE)) +
    scale_color_brewer(palette = "Dark2")

```